Hdf5 application

msandre opened the pull request:
For reading and writing a model part / results in hdf5.

jcotela committed:
Initial commit of HDF5 application.

jcotela committed:
Defining WriteModelPart function.

jcotela committed:
Minimal support to write a model part in hdf5.

Commit:
fixed merge conflict in cmake

Commit:
added class HDF5File.

Commit:
switched to link to hdf5 c lib.

Commit:
fixed compile errors for HDF5File.

Commit:
added some hdf5 tests, fixed some (not all) bugs.

Commit:
serial tests working and code cleaned up.

Commit:
added checks for serial/parallel hdf5 to cmake.

Commit:
starting point for mpio with hdf5 file.

Commit:
refactored/cleaned path manipulation functions.

Commit:
small.

Commit:
small cleanup.

Commit:
adding hdf5 file mpi to compilation.

Commit:
added mpi link libs for parallel hdf5.

Commit:
removed GetRank

Commit:
changed hdf5 read interface.

Commit:
small.

Commit:
small.

Commit:
added independent/collective data access.

Commit:
making domains graph const.

Commit:
cleaned parallel fill communicator.

Commit:
small.

Commit:
adjusted writeimpl for mpi with independent mode.

Commit:
fix comment.

Commit:
changed default to collective io.

Commit:
made base hdf5 file members private.

Commit:
fixed compiler errors.

Commit:
added class docstrings.

Commit:
added write partition.

Commit:
added hdf5 model part io definition.

Commit:
added GetPID.

Commit:
added GetTotalProcesses to hdf5 file.

Commit:
small.

Commit:
comments.

Commit:
added serial hdf5 read/write nodes.

Commit:
added parallel hdf5 read/write of nodes.

Commit:
small.

Commit:
changed to ublas vector for interface of hdf5 file.

Commit:
added separate interface base class for hdf5 file.

Commit:
added matrix read/write.

Commit:
added matrix test.

Commit:
added matrix io.

Commit:
added serial write elements.

Commit:
switched to typeid for element checking.

Commit:
added write conditions.

Commit:
added read elements.

Commit:
added read conditions serial.

Commit:
omp loop.

Commit:
removed read connectivities.

Commit:
added write attribute.

Commit:
started get attribute.

Commit:
added read attribute.

Commit:
finishing hdf5file serial tests.

Commit:
added H5close to tests.

Commit:
added mem leak check for hdf5.

Commit:
moved data type identification to utils.

Commit:
small.

Commit:
added hdf namespace and mesh io.

Commit:
cleaning serial model part io.

Commit:
switched to raw ptrs for SetData().

Commit:
added bins utility.

Commit:
small.

Commit:
added omp in bin creation.

Commit:
made bin creation faster for single element type.

Commit:
orgainizing/cleaning includes.

Commit:
added io functions for partitioned model part.

Commit:
added modelpartio tests and fixed binutil bug.

Commit:
added begin() end() methods to data value container.

Commit:
added write properties.

Commit:
small.

Commit:
added properties io.

Commit:
small.

Commit:
properties io impl.

Commit:
fixed/tested properties io.

Commit:
small.

Commit:
small.

Commit:
added variables list io.

Commit:
implemented nodal solution step data io.

Commit:
cleaned utils.

Commit:
small.

Commit:
small

Commit:
added partitioned parameter.

Commit:
small.

Commit:
added const method.

Commit:
fixed bug in nodes variable list.

Commit:
added python test for model part io.

Commit:
added python test for hdf5 nodal solution data io.

Commit:
small.

Commit:
added process info io.

Commit:
small.

Commit:
small.

Commit:
updating partitioned io.

Commit:
small.

Commit:
small

Commit:
improved write error message

Commit:
added partition index io.

Commit:
added test for partitioned model part io.

Commit:
added io timers.

Commit:
cleaned timer output in parallel io.

Commit:
added nodal solution data parallel io test.

Commit:
small.

Commit:
add path to error msg.

Commit:
added python cpp unit test script.

Commit:
merged master to hdf5application.

Commit:
merged master (2).

Commit:
updated license.

Commit:
updated license.

msandre commented:
There appears to be a problem after merging with recent master branch. Line HDF5Application/custom_io/hdf5_nodal_solution_step_variables_io.cpp:55 where VariablesList::Add() is called appears to through seg fault at kratos/containers/variables_list.h:398. I will try to fix it, but  if anyone who knows about recent changes in variables_list.h might have an idea please post.

Commit:
fixed invalid object bug in variables list clear().

msandre commented:
Problem is fixed. See #998 .

jcotela committed:
Changing HDF5 version check.

jcotela commented:
Hi Mike, I'm trying to test this branch. I've had to change the CMake version check for HDF5, since I have version 1.10 and the old one was checking for exactly 1.8.

By the way, have you had problems with the zlib library? In my local setup, I have clashes between the system zlib (used by hdf5) and a custom zlib we compile for gidpost. I have done a dirty hack on my machine for now, but I'll try to push a fix on this branch later.

msandre commented:
Hey Jordi, I didn't test for version 1.10 but I think 1.10 has a small number of differences in a few interface functions (number of function args if I understood correct). As long as it compiles I think all the functions I am using are the same in 1.10.

I didn't have a problem with zlib but there was the following cmake warning:
```
CMake Warning at applications/HDF5Application/CMakeLists.txt:85 (add_library):
  Cannot generate a safe runtime search path for target KratosHDF5Application
  because files in some directories may conflict with libraries in implicit
  directories:

    runtime library [libz.so.1] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /home/mike/Kratos/Kratos/cmake_build/external_libraries/zlib

  Some of these libraries may not be found correctly.
```
I don't have this folder on my ld path (I'm using only rpath) so maybe this is why. I didn't completely understand the warning but we will have to think of a solution for this.

jcotela commented:
Ok, in this case I'll just go ahead and test it with the 1.10 version. Hopefully it won't change that much, but it is the one that comes by default on my laptop (manjaro linux) and I'd prefer to avoid installing a custom hdf5 if possible.

Regarding the CMake zlib issue, it is the same one for me, but it is "fatal" for my cmake for some reason... I'll investigate it a bit more, but we'll need some kind of fix. The problem happens because gidpost uses zlib too, which we compile ourselves (because it is never present in windows). This custom library clashes with the "system" zlib used by hdf5. I'm trying to only compile our own zlib if none is found in the system, but I haven't been able to get cmake to do this so far...

jcotela commented:
Hey Mike, I'm running the HDF5 application with hdf5 1.10, and I think it works. I'm getting some "unexpected" output on the console for several of the tests, things like:

```
HDF5-DIAG: Error detected in HDF5 (1.10.1) thread 0:
  #000: H5F.c line 769 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet7                                                OK.
```

I understand that this is normal (since the tests themselves expect an exception to be thrown). However, can you confirm that this is what you would expect to see? I'd like to be sure that this is not a "bug" introduced by the version upgrade. Thanks!

I'm also working on a fix for the zlib issue but, since it won't have any effect on this branch (other than removing the warning), I'll address it later on a separate PR.

msandre commented:
Yes, that's the same as for hdf5 1.8. As I was programming the hdf5 part I didn't see any mention of differences in 1.10 for the functions I used. hdf1.8 was the version available on stock ubuntu 16 so that's why I used it and added the check on the version since I hadn't tested any other version. Thanks for checking on 1.10.

jcotela commented:
Hi Mike, I'm trying to play a bit with the hdf5 i/o. How should I use the output facilities? I'm running into some problems with the following code (at the end of the "standard" python script for the fluid solver):
```
h5_file_parameters = Parameters(r'''{
    "file_name" : "hdf5_test"
}''')
h5_io_parameters = Parameters(r'''{}''')
h5_out_file = HDF5File(h5_file_parameters)

h5_io = HDF5ModelPartIO(h5_io_parameters,h5_out_file)
h5_io.WriteModelPart(main_model_part)
```

which when writting the buffer size (?)
```
RuntimeError: Error: Invalid path: 



in /home/jcotela/Kratos/applications/HDF5Application/custom_io/hdf5_file.cpp:173:void HDF5::File::AddPath(__cxx11::string)
   /home/jcotela/Kratos/applications/HDF5Application/custom_io/hdf5_nodal_solution_step_variables_io.cpp:82:void HDF5::Detail::NodalSolutionStepVariablesIO::WriteBufferSize(int)
   /home/jcotela/Kratos/applications/HDF5Application/custom_io/hdf5_model_part_io.cpp:232:virtual void HDF5::ModelPartIO::WriteModelPart(ModelPart&)

```

I'm I doing this right or should I try something different?

msandre commented:
It's an oversight on my part. In your case, the string is empty and the path is considered invalid (should start with /). I will fix this, in the meantime it should work if you just set the prefix to the model part IO to something like /ModelData. You may also need to write hdf5_test.h5 instead of hdf5_test.

jcotela commented:
Thanks! That solved it. Unfortunately I get a segfault now... I'll report back once I understand why.

Commit:
fix empty prefix parameter.

Commit:
Merge branch 'experimental-hdf5-support2' of https://github.com/KratosMultiphysics/Kratos into experimental-hdf5-support2

msandre commented:
Ok, the prefix problem should be fixed now.

jcotela commented:
My segfault was caused by not passing the element_list and condition_list parameters to the IO, it seems to work now. Let me look around a bit more and I'll give a review (sorry about the delay).

jcotela commented on applications/HDF5Application/CMakeLists.txt:
```diff
@@ -0,0 +1,106 @@
+  endif()
+else()
+  if(HDF5_IS_PARALLEL)
+    message(FATAL_ERROR "Parallel HDF5 libraries were found.")
```
I encountered this message while trying to compile the application and I found it slightly confusing... Maybe specify that this is a non-parallel compilation, so it needs a non-parallel hdf5 version?

msandre replied:
fixed.

jcotela commented on applications/HDF5Application/custom_python/add_custom_io_to_python.cpp:
```diff
@@ -0,0 +1,90 @@
+{
+    using namespace boost::python;
+
+    class_<HDF5::File, HDF5::File::Pointer, boost::noncopyable >("HDF5File", init<Parameters&>())
```
If I understand this correctly, this class is not useful in practice, it just defines the python interface for derived classes. May I suggest passing `no_init` as the constructor argument, so that it cannot be constructed from Python?

msandre replied:
fixed.

jcotela commented on applications/HDF5Application/custom_python/add_custom_io_to_python.cpp:
```diff
@@ -0,0 +1,90 @@
+    .def("GetFileName",&HDF5::File::GetFileName)
+    ;
+
+    class_<HDF5::FileSerial, HDF5::FileSerial::Pointer, bases<HDF5::File>, boost::noncopyable>(
```
I understand why you placed everything on a different namespace, but it generates some confusion with the python interface, since this one does not follow the same convention and just prepends `HDF5` to class names. What would you think of adding all these functions to an `hdf5` module in Python to replicate the same structure?

msandre replied:
Do you mean in addition to the KratosHDF5Application to have something like a KratosHDF5 module and access the classes like KratosHDF5.SerialFile()? I'm not sure how this can be done. I find the boost documentation on the python interface quite limited. Do you know an example or some reference that does something like what you have in mind? Alternatively, I could just remove the HDF5 prefix from the python class names and use the single module as a namespace. Would this be better?

jcotela replied:
I see two possibilities here: either we rename the boost python module (which can be done but has some annoying details I don't remember right now) or we do something custom in the HDF5Application.py to place these functions in a different namespace. In any case, I can take care of that if we decide we want to do it like this.
In any case, I prefer to keep it as it is now rather than using only the module name (I think this is a bit error-prone, since we are not so consistent in importing applications with relative scopes in kratos).

Maybe this is not so important right now, it is just a discussion I wanted to have.

msandre replied:
ok, I think it's good to keep in mind, but I will not modify it here.

jcotela commented on applications/HDF5Application/custom_io/hdf5_model_part_io.cpp:
```diff
@@ -0,0 +1,265 @@
+            "prefix": "",
+            "list_of_elements": [],
+            "list_of_conditions": []
+        })");
```
these `list` arguments are a bit unfortunate, since the mdpa should already have this information, just not in a convenient way (and I guess the hdf5 file input too).

What is the rationale for adding them? I would be in favor of parsing the input file once in advance to construct these lists if required.

msandre replied:
Initially the hdf5 input file doesn't exist. The problem I encountered was how to identify the element type in Kratos (these may not even be given in the mdpa). There is no function that returns the element's type (e.g., by string). If there were such a function, I could loop over all elements and conditions and identify the different types in the model part, but this would add computational overhead in exchange for user convenience. It would nevertheless simplify the function IsMatch() in the PointerBinsUtility. Currently, typeid by itself fails when the geometry is different but the element class is the same so I also have to check number of nodes and working space dimension.

jcotela replied:
Ok, understood. You are right that the mdpa type is not necessarily the one you want. I would leave it as is for now, since there is no clean solution. However I think there is need in kratos of a functionality like you describe.

Thinking ahead, if I were to use the file you produce with hdf5 as input for a new simulation (say like a restart), could I avoid having to provide this string? I think having a way to do this will help us in the future.

Also, this is more of a wish, but one thing we want in the future is the ability to read pure geometries (not elements) and generate the elements depending on the solver. Unfortunaly we have no container for geometries on the model part, so it can't be done now. My question is: given the implementation of a container for lists of geometries, do you see your I/O as being able to work in this way?

msandre replied:
It is possible to automatically extract the element and condition types from the file directly for reading a model part. In python this can be done quite easily, in c++ we need to add the functionality to the file class. Currently, if you don't have the original json parameters you can also view the file hierarchy with hdfview and manually create the lists.

To your last question, I think it should be straightforward to extend. I think the bigger challenge is managing all the possible formats for io. Probably, it is easier to write a new class in this case. As it is now, there is not a lot of code in the model_part_io. Hopefully this makes it easier to write new types of io without having to worry so much about the low level stuff.

pooyan-dadvand replied:
If you need the list of all element names already registered (like "TotalLagrangian3D4N") you can make a loop over `KratosComponents<Element>` and take the `it->first` which is the name of the element. The same can be done for `KratosComponents<Condition>`

jcotela reviewed (commented):
Finally a review, sorry for the delay...

First, great work! The lower-level stuff is well organized and with lots of error-checking. Also, :+1: for the extensive testing.

I must admit I have been unable to produce MPI output though... I'm sure I'm initializing something wrong, but it is not working (I'll post the test code later, since I have it on a different computer).

I do have a few questions/comments. Most of them are in the review, except for this one (didn't know exactly where to put it):

The way this I/O class works is different from the mdpa I/O in one detail: if the output file exists, it is not overwritten. This forces us to delete old output before running new simulations, which is a bit annoying when testing (I'm running something, which fails, now I have to delete incomplete output before trying again...). I know it is a minor inconvenience, but I wanted to raise the issue to see if there is a technical reason for this.

jcotela commented:
My test code for MPI:

```
h5_file_parameters = Parameters(r'''{
    "file_name" : "hdf5_test.h5"
}''')
h5_file_parameters["file_name"].SetString("hdf5_test_{0}.h5".format(mpi.rank))
h5_io_parameters = Parameters(r'''{
    "prefix" : "/ModelData",
    "list_of_elements" : ["VMS2D"],
    "list_of_conditions" : ["WallCondition2D2N"]
}''')
h5_out_file = HDF5FileParallel(h5_file_parameters)

h5_io = HDF5PartitionedModelPartIO(h5_io_parameters,h5_out_file)
h5_io.WriteModelPart(main_model_part)
```

The error: (running with 2 threads, both output the same message)
```
HDF5-DIAG: Error detected in HDF5 (1.10.1) HDF5-DIAG: Error detected in HDF5 (1.10.1) MPI-process 1:
  #000: H5Dio.c line 269 in H5Dwrite(): can't prepare for writing data
    major: Dataset
    minor: Write failed
  #001: H5Dio.c line 345 in H5D__pre_write(): can't write data
    major: Dataset
    minor: Write failed
  #002: H5Dio.c line 676 in H5D__write(): collective access for MPI-based driver only
    major: Dataset
    minor: Feature is unsupported
MPI-process 0:
  #000: H5Dio.c line 269 in H5Dwrite(): can't prepare for writing data
    major: Dataset
    minor: Write failed
  #001: H5Dio.c line 345 in H5D__pre_write(): can't write data
    major: Dataset
    minor: Write failed
  #002: H5Dio.c line 676 in H5D__write(): collective access for MPI-based driver only
    major: Dataset
    minor: Feature is unsupported
Traceback (most recent call last):
  File "MPIKratos.py", line 185, in <module>
    h5_io.WriteModelPart(main_model_part)
RuntimeError: Error: H5Dwrite failed. Please ensure global data set is non-empty.
Path: "/ModelData/Nodes/Local/Ids".
``` 

Could this be a version-incompatibility issue?

jcotela commented:
@msandre I notice the branch has some merge conflicts. In the case of `applications/applications_interface.py`, please consider that this file implements the obsolete application importing mechanism and is there only for backwards compatibility (no need to add you application there)

Commit:
fixed error message in cmake.

Commit:
make python file no_init.

msandre commented:
It is possible to truncate the file by setting the json parameter `"file_access_mode"` to `"truncate"`. By default though it throws an error if the file exists. I did this to be on the safe side that a user knowingly overwrites his or her data.

msandre commented:
I'm not sure why the parallel simulation fails. Could you upload the test case? When I was testing, the only time I observed a failure in this function was when there was nothing to write (i.e. the global data set was empty). I would like to see if your example works on my machine.

Thanks for the review.

Commit:
undid changes to applications_interface.py

msandre commented:
Just another thought on the failing mpi. At the moment there is no IO for submodelparts. Maybe this is creating a problem for your case. I decided to leave the submodelpart io for a later pull request since it is already a lot of code.

pooyan-dadvand reviewed (changes requested):
I have started to hover over your PR and it is really big! I can tell you that from hovering point of view your implementation is good as usual. Nevertheless, I got lost in understanding your design layers and object roles. It would be great if you can give us an overview of your design (small class diagram with role description would be perfect) This also helps me to see which parts can be injected directly to the core. Also I wanted to mention:
1. I liked the use of HDF5 namespace which is something new and can be used in other places to improve code organization
2. It would be great if you change the Detail namespace to Internals so it is compatible with the rest of the code?

msandre commented:
I tried sketching out the classes in the attachment. Let me know if something is not clear.

[2406_001.pdf](https://github.com/KratosMultiphysics/Kratos/files/1514492/2406_001.pdf)

I will change Detail to Internals as soon as I'm back on the branch.

Commit:
changed Detail to Internals.

pooyan-dadvand commented:
@msandre Thank you very much for the sketch! It was really helpful! Now I see your global view which helps a lot!

Understanding your design I see that your `File` class is actually HDF5 concern class and not something to be reused in core. 

Now my main concern is the `PartitionedModelPartIO` which is not based on `ModelPartIO`. I have checked your implementation and seems that you had to duplicate everything because you are mixing the partition information (local, ghost) with usual data (coordinates, connectivity) Which reduces the modularity of  your input. Is there any strong reason to do that?

msandre commented:
Looking back on it, I considered 3 possibilities: 1) Make `PartitionedModelPartIO` and `ModelPartIO` separate classes derived from `IO`, like I did. 2) Make them one class with a parameter `partitioned`. 3) Derive `PartitionedModelPartIO` from `ModelPartIO`. My choice for not doing 3) was to keep the class hierarchy simple and avoid too much inheritance (e.g., mixing variables in base and derived classes, having to add more constructors). I also imagined there could be more classes in the future and it might be best to have a flat hierarchy with implementation details encapsulated in separate helper classes. At the time, I decided against 2) mainly to make the purpose of the classes more explicit through their separate names and avoid lots of if statements / code complexity. I realize this is not the same as I did for the solution data IO, but in that case it seemed to be less complicated. I didn't understand your comment about the modularity of the input. Could you explain?

pooyan-dadvand commented:
I agree that the option 2 is not the best choice. So let's focus on 1 and 3. I see  that you didn't want to derive to make a flat hierarchy. Nevertheless the hierarchy here is not very large (only 2 layers which might be extended to 3) On the other hand deciding to avoid deriving should result in delegation and NOT duplication. This means that in such case the partitioned IO should have a member which is the normal one and delegate the work to it when is possible. For maintaining the encapsulation in a hierarchy a standard way would be to use only public interface of the base class. So the implementations are decoupled in a same way that with any other classes.

Saying that I would have gone for option 3 due to its simplicity considering the following conditions: 
 - Two classes should have almost the same interface
 - There are only two related classes 

In case that you see more classes in the future we might go for another solution like [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern). 
 -  Then you should derive an `ExtendedHDF5IO` from your normal one which is the common denominator and make `ExtendedHDF5IO` have normal `HDF5IO` as member and delegating all its methods to that. 
 - Deriving each extension (like `PartitionedHDF5IO`) from the extended and add the extra capabilities. 
 - The advantage here is that if somebody create another IO derived from `HDF5IO` you can extend it with your extensions without effort. 

Meanwhile, this implies that you should divide your input in standard and extended part to be easier the IO implementation and less duplication. This is what I was referring by input modularity.

Commit:
restructured model part io.

Commit:
added tests to smallsuite.

Commit:
added write model part to base.

msandre commented:
Ok, I would prefer that PartitionedModelPartIO does not inherit from ModelPartIO, so I added a common implementation base class ModelPartIOBaseImpl. I also gutted the read/write functions for conditions and elements and put the implementation in connectivities io. @pooyan-dadvand @jcotela please tell me if there are more changes needed.

Thanks!

Commit:
moved destructor  to protected.

pooyan-dadvand reviewed (dismissed):
Thanks for your changes, now I see it more readable and I liked the use of internals which hides the implementation details. 

Just one minor suggestion, would you please leave the base class name without final impl? It is not true as the implementation is in internals classes....

Commit:
removed impl from base class name.

Commit:
merged master.

Commit:
fixed constructor.

msandre commented:
@pooyan-dadvand I removed the impl like you said. I must have missed some feature on GitHub, because it says I dismissed your review after I added the new commit (and merged master). Can you approve it once more?

Thanks

pooyan-dadvand reviewed (approved):
Recently we have activated the setting which needs re-approval of the PR after any new change. This has been made after some cases in which developers after approval continued working on same branch by mistake and uploaded some erroneous code to master.

msandre commented:
Thanks, @jcotela is it alright to delete the experimental-hdf5-support branch?

jcotela commented:
@msandre I've gone ahead and deleted it. Thanks for developing the HDF5 support.
